library/aes.c

```diff
@@ -1506,6 +1606,69 @@ int mbedtls_aes_self_test( int verbose )
         mbedtls_printf( "\n" );
 #endif /* MBEDTLS_CIPHER_MODE_CFB */
 
+#if defined(MBEDTLS_CIPHER_MODE_OFB)
+    /*
+     * OFB mode
+     */
+    for( i = 0; i < 6; i++ )
+    {
+        u = i >> 1;
+        keybits = 128 + u * 64;
+        mode = i & 1;
```

Comment:
We don't need to test both encrypt and decrypt, since there is no `mode` parameter given to `mbedtls_aes_crypt_ofb`.

Reply:
The NIST test parameters specify both encryption and decryption, and `mode` is used to know the direction to handle the buffers - plaintext to ciphertext or vice versa.

If we remove `mode`, we will compromise the testing by only testing encryption, which isn't what's specified in NIST. You could argue that OFB is symmetrical, and we only need to test encryption, but personally I'd prefer to say we do the NIST test vectors fully, without qualification.

We could structure the loop in a different way, with a different `for` loop and different logic, but I think this is a fair compromise, it works and I don't think it's worth changing.

Reply:
I agree with Simon. We know the abstract OFB function is symmetrical, but we don't know our implementation has that property too until we test it. I think tests should assume as few prior knowledge as reasonably possible (and in that case, the extra cost of testing both ways is close to 0).

We test other symmetric modes (as well as stream ciphers) both ways too, and I think that's the right thing to do.

Reply:
For the self test, I'm not sure we need to test both ways; I could see benefit in just checking encryption for the self test, as it gives the same amount of code coverage and potentially saves time when a system boots up. I was would not suggest to removing the encrypt and decrypt tests from `test_suite_aes`.

Reply:
If you're really keen on optimising the tests, I suggest you look at the test vectors and determine why those values were chosen, what they are exercising, and whether the decryption  test vectors are testing 'decryption', or a set of values chosen to exercise the function.

I personally think this is a bad idea because you're assuming you know better than NIST, in what test vectors to execute and are effectively cherry-picking them.